File: ahlQuran_Back/app/schemas/achievement.py
from pydantic import BaseModel, Field
from datetime import datetime
from typing import Optional
from enum import Enum


class AchievementType(str, Enum):
    NORMAL = "normal"
    SMALL = "small"
    BIG = "big"
    
    def __str__(self):
        return self.value


class AchievementBase(BaseModel):
    from_surah: int = Field(..., ge=1, le=114, example=2)  # Chapter number 1-114
    to_surah: int = Field(..., ge=1, le=114, example=2)    # Chapter number 1-114
    from_verse: int = Field(..., gt=0, example=1)
    to_verse: int = Field(..., gt=0, example=10)
    note: Optional[str] = Field(None, example="Good memorization")
    achievement_type: AchievementType = Field(default=AchievementType.NORMAL, example="normal")
    date: str = Field(..., description="Date in DD-MM-YYYY format", example="19-12-2024")


class AchievementCreate(AchievementBase):
    student_id: int = Field(..., gt=0, example=1)


class AchievementUpdate(BaseModel):
    from_surah: Optional[int] = Field(None, ge=1, le=114)  # Chapter number 1-114
    to_surah: Optional[int] = Field(None, ge=1, le=114)    # Chapter number 1-114
    from_verse: Optional[int] = Field(None, gt=0)
    to_verse: Optional[int] = Field(None, gt=0)
    note: Optional[str] = None
    achievement_type: Optional[AchievementType] = None
    date: Optional[str] = Field(None, description="Date in DD-MM-YYYY format")


class AchievementResponse(AchievementBase):
    id: int
    student_id: int
    created_at: datetime
    updated_at: Optional[datetime]

    class Config:
        from_attributes = True


class AchievementList(BaseModel):
    achievements: list[AchievementResponse]
    total: int

File: ahlQuran_Front/lib/system/new_models/achievement.dart
import '../../../data/quran_data.dart';

class Achievement {
  final int? id;
  final int studentId;
  final int fromSurah; // Changed to chapter number
  final int toSurah; // Changed to chapter number
  final int fromVerse;
  final int toVerse;
  final String? note;
  final String achievementType; // "normal", "small", "big"
  final String date; // Format: DD-MM-YYYY
  final DateTime? createdAt;
  final DateTime? updatedAt;

  Achievement({
    this.id,
    required this.studentId,
    required this.fromSurah,
    required this.toSurah,
    required this.fromVerse,
    required this.toVerse,
    this.note,
    this.achievementType = 'normal',
    required this.date,
    this.createdAt,
    this.updatedAt,
  });

  factory Achievement.fromJson(Map<String, dynamic> json) {
    return Achievement(
      id: json['id'],
      studentId: json['student_id'],
      fromSurah: json['from_surah'],
      toSurah: json['to_surah'],
      fromVerse: json['from_verse'],
      toVerse: json['to_verse'],
      note: json['note'],
      achievementType: json['achievement_type'] ?? 'normal',
      date: json['date'] ?? '',
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'])
          : null,
      updatedAt: json['updated_at'] != null
          ? DateTime.parse(json['updated_at'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) 'id': id,
      'student_id': studentId,
      'from_surah': fromSurah,
      'to_surah': toSurah,
      'from_verse': fromVerse,
      'to_verse': toVerse,
      'note': note,
      'achievement_type': achievementType,
      'date': date,
      if (createdAt != null) 'created_at': createdAt!.toIso8601String(),
      if (updatedAt != null) 'updated_at': updatedAt!.toIso8601String(),
    };
  }

  Achievement copyWith({
    int? id,
    int? studentId,
    int? fromSurah,
    int? toSurah,
    int? fromVerse,
    int? toVerse,
    String? note,
    String? achievementType,
    String? date,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Achievement(
      id: id ?? this.id,
      studentId: studentId ?? this.studentId,
      fromSurah: fromSurah ?? this.fromSurah,
      toSurah: toSurah ?? this.toSurah,
      fromVerse: fromVerse ?? this.fromVerse,
      toVerse: toVerse ?? this.toVerse,
      note: note ?? this.note,
      achievementType: achievementType ?? this.achievementType,
      date: date ?? this.date,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  // Helper method to get surah name
  String getFromSurahName() {
    final surah = getSurahByNumber(fromSurah);
    return surah?.name ?? 'سورة $fromSurah';
  }

  String getToSurahName() {
    final surah = getSurahByNumber(toSurah);
    return surah?.name ?? 'سورة $toSurah';
  }
}
